WordPress/gutenberg
# Components: add `Tabs` (a composable `TabPanel` v2)

Addresses #52997

## What?
Add a new `Tabs` component that will serve as a v2 replacement of the current `TabPanel`.

## Why?
The new component will be more composable, allowing consumers to tailor their implementations to meet their needs more easily than can currently be accomplished. We'll also be looking to implement some previously requested features/enhancements with this new component as well.

## How?
The main `Tabs` component will have three subcomponents, each handling their respective `Ariakit` counterparts: `Tabs.TabList`, `Tabs.Tab`, and `Tabs.TabPanel`. The `Tabs` component will house the `TabList` (which will, in turn, contain the various `Tab`s) and the individual `TabPanel`s. Each `Tab` will correspond to one `TabPanel`:

```
<Tabs { ...props }>
	<Tabs.TabList>
		<Tabs.Tab id={ 'tab1' } title={ 'Tab 1' }>
			Tab 1
		</Tabs.Tab>
		<Tabs.Tab id={ 'tab2' } title={ 'Tab 2' }>
			Tab 2
		</Tabs.Tab>
		<Tabs.Tab id={ 'tab3' } title={ 'Tab 3' }>
			Tab 3
		</Tabs.Tab>
	</Tabs.TabList>
	<Tabs.TabPanel id={ 'tab1' }>
		<p>Selected tab: Tab 1</p>
	</Tabs.TabPanel>
	<Tabs.TabPanel id={ 'tab2' }>
		<p>Selected tab: Tab 2</p>
	</Tabs.TabPanel>
	<Tabs.TabPanel id={ 'tab3' }>
		<p>Selected tab: Tab 3</p>
	</Tabs.TabPanel>
</Tabs>
```

Thus far, the work on this composable `Tabs` component has focused on creating the subcomponents and ensuring they function in a way that maintains parity with `TabPanel`, and only making intentional changes. Focuses so far have been:
- Implementing the subcomponents
- Ensuring `TabPanel`'s stories are replicated and functional in Storybook
- Ensuring `TabPanel`'s unit tests are updated and passing for `Tabs`. There were a few small changes needed here, one of which regarding when the `onSelect` callback is triggered that I'll touch on more below.

Remaining todos include:
- [x] [Controlled Mode](https://github.com/WordPress/gutenberg/issues/27371). Specifically, this will allow consumers to set the selected tab when necessary.
- [x] The ability to place the `tablist` and `tabpanel` subcomponents in arbitrary locations within the DOM.
- [x] `tablist` design flexibilty
  - Placing an auxiliary button (or other element?) such as a "Close" button beside the tablist
  - Optional https://github.com/WordPress/gutenberg/pull/52845
- [x] Limit `onSelect` calls to intentional user selections. Currently in `TabPanel`, the consumer-provided `onSelect` callback is triggered when the initial tab is selected during the first render. In `Tabs`, we intend to only trigger this callback when the user actively selects a tab, not when the component does so automatically.
- [x] The usual README and CHANGELOG goodness
- [x] MAYBE/NICE TO HAVE: tab stop/focus behavior around `tabpanels` with focusable child elements. See #52997 for a more detailed description.

## Notes/Thoughts/Questions
- Currently, we have everything in a wrapper component (`Tabs`) which supports slotfills. `Tabs` also handles providing context to the different subcomponents. Are there potential or existing implementations where a wrapper would be problematic? If so, we may want to look into alternative approaches that don't require a single wrapper component and context provider.
- I've noticed some interesting `Ariakit` behavior that I wanted to get @diegohaz's feedback on, if you have a chance. It looks like the `setSelectedId` callback (analogous to our component's `onSelect`) is not called when `Ariakit.useTabStore()` receives a `defaultSelectedId`. It _is_ called when a `defaultSelectedId` is _not_ present. Does that sound like the expected behavior? I suspect it's because without that prop `selectedId` is `undefined` and needs to be updated, triggering the callback, but I wanted to double check. If that's what's happening, I might try defaulting `defaultSelectedId` to the first enabled tab without our implementation.


## Testing Instructions
- Unit tests pass
- Launch Storybook
- Confirm the tests that are shared between `Tabs` and `TabPanel` behave the same in both components. There may be some minor differences (for example, I added a third tab on some for `Tabs`), but the behavior should be the same, even if the tabs themselves differ slightly.
- When testing the shared stories, confirm the behavior is consistent for both keyboard and mouse interactions.
- On the `Using SlotFill` story:
    - Confirm that the `tabpanel` content updates properly when changing tabs, even though it's being rendered inside a different DOM element.
    - Confirm keyboard navigation (ie, it should still be possible to tab between the `tablist` and the `tabpanel`, while arrow keys navigate between the individual tabs themselves.
- Nothing specific to test on the `Insert Custom Elements` story, but give it a go and try to break stuff 😉
- On the `Controlled Mode` story:
    - Test mouse and keyboard interactions on the tabs, confirm they behave the same way other stories do.
    - Test the provided dropdown and ensure it allows you to change tabs
    - Test a mixed approach, switching between the tabs and the dropdown, and confirm nothing gets out of sync. For example, use the dropdown to get to a specific tab, and then manually click a new one. Can you now use the dropdown to get _back_ to the original tab again?
- Style/design flexibility:
    - Each of the new subcomponents has a `style` prop that can be used to give consumers more control over the look and feel of the `tabpanel` when needed (for example the 'fitted tabs' pr referenced above). Try modifying the default `Tabs` story by adding style props to the different components, and confirm your styles are added and rendered correctly.


## Follow-ups

- We should add the `Tabs` component to the package's private API (lock/unlock), and start using it around the editor. This will be a great way to test these components in a real-world scenario. We could start from those instances where folks had to be a bit hacky about adding a close button at the end of the tablist
- Once we're happy with `Tabs`, we should:
  - Refactor all usages of `TabPanel` to `Tabs` in the project
  - Export the component from the package as a stable component (ie. not through private APIs)
  - Mark the `TabPanel` component as deprecated
